edge/controller/env/broker.go

```diff
@@ -513,58 +448,42 @@ func (b *Broker) sessionCreateEventHandler(args ...interface{}) {
-	addByClusterId := map[string]*edge_ctrl_pb.SessionAdded{}
+func (b *Broker) sendSessionCreates(session *persistence.Session) {
+	sessionAdded := &edge_ctrl_pb.SessionAdded{}
 
-	for _, session := range sessions {
-		service, err := b.ae.Handlers.Service.HandleRead(session.ServiceId)
-		if err != nil {
-			log := pfxlog.Logger()
-			log.WithField("cause", err).Error("could not send network session added, could not find service")
-			continue
-		}
-
-		for _, clusterId := range service.Clusters {
-			var ok bool
-			if _, ok = addByClusterId[clusterId]; !ok {
-				addByClusterId[clusterId] = &edge_ctrl_pb.SessionAdded{}
-			}
+	service, err := b.ae.Handlers.Service.HandleRead(session.ServiceId)
+	if err != nil {
+		log := pfxlog.Logger()
+		log.WithField("cause", err).Error("could not send network session added, could not find service")
+		return
+	}
 
-			fps, err := b.getActiveFingerprints(session.Id)
+	fps, err := b.getActiveFingerprints(session.Id)
 
-			if err != nil {
-				pfxlog.Logger().Errorf("could not obtain a fingerprint for the api session [%s] and session [%s]", session.ApiSessionId, session.Id)
-				continue
-			}
-
-			svc, err := b.modelServiceToProto(service)
+	if err != nil {
+		pfxlog.Logger().Errorf("could not obtain a fingerprint for the api session [%s] and session [%s]", session.ApiSessionId, session.Id)
+		return
+	}
 
-			if err != nil {
-				pfxlog.Logger().Errorf("could not convert service [%s] to proto: %s", service.Id, err)
-				continue
-			}
+	svc, err := b.modelServiceToProto(service)
 
-			addByClusterId[clusterId].Sessions = append(addByClusterId[clusterId].Sessions, &edge_ctrl_pb.Session{
-				Token:            session.Token,
-				Service:          svc,
-				CertFingerprints: fps,
-				Hosting:          session.IsHosting,
-			})
-		}
+	if err != nil {
+		pfxlog.Logger().Errorf("could not convert service [%s] to proto: %s", service.Id, err)
+		return
 	}
 
-	for clusterId, nsr := range addByClusterId {
-		entries := b.clusterMap.GetOnlineEntries(clusterId)
-		for _, e := range entries {
+	sessionAdded.Sessions = append(sessionAdded.Sessions, &edge_ctrl_pb.Session{
+		Token:            session.Token,
+		Service:          svc,
+		CertFingerprints: fps,
+		Hosting:          session.IsHosting,
+	})
 
-			if buf, err := proto.Marshal(nsr); err == nil {
-				if err = e.Channel.Send(channel2.NewMessage(SessionAddedType, buf)); err != nil {
-					pfxlog.Logger().WithField("cause", err).Error("error sending network session added")
-				}
-			} else {
-				pfxlog.Logger().WithField("cause", err).Error("error sending network session added, could not marshal message content")
-			}
-		}
+	if buf, err := proto.Marshal(sessionAdded); err == nil {
+		msg := channel2.NewMessage(SessionAddedType, buf)
+		b.sendToAllEdgeRoutersForSession(session.Id, msg)
+	} else {
+		pfxlog.Logger().WithField("cause", err).Error("error sending network session added, could not marshal message content")
```

Comment:
Logging messages in the format of `WithField("cause", err)` we should change them to `WithError(err)`. `WithError()` is simply `WithField("error", err)` but it more semantic and most of the other ziti codebase does it that way.

I simply didn't know about WithError till it was too late.

Reply:
:+1: Fixed.

Reply:
I actually prefer `logrus.Errorf("message (%w)", err)`-style wrapped errors. I find both the messages, and the code easier to read.

Reply:
I'll check it out. I'm dissatisfied with the current state of error handling in general. I've been using `Errorf("%+v", err)` in order to view stack traces when they are available. Will %w do that as well?

Reply:
Wrapping errors (so that they can be unwrapped later) is the primary reason for using `%w`. I believe the point of wrapping is so they can be unwrapped later. That may not help with stack traces. 

The `ziti-fabric` codebase still uses `%s` in a lot of places, but I plan on migrating them over to `%w` as I work through there.